pkg/generate/code/compare_helpers.go

```diff
@@ -0,0 +1,460 @@
+				out += equalSlice(
+					shape,
+					firstFuncParam,
+					secondFuncParam,
+					1,
+				)
+			default:
+				msg := "Should not generate a helper function for any type different" +
+					"than string to struct/list/map map and struct/list/map slices"
+				panic(msg)
+			}
+
+			// return true
+			out += fmt.Sprintf("\treturn true\n")
+			// }
+			out += "}\n"
+		}
+	}
+	return out
+}
+
+// expectHelperFunction returns true if a given shape needs a helper function, false
+// otherwise.
+func expectHelperFunction(
+	shape *awssdkmodel.Shape,
+) bool {
+	if shape.Type == "map" {
+		valueType := shape.ValueRef.Shape.Type
+		return valueType == "structure" || valueType == "list" || valueType == "map"
+	}
+
+	if shape.Type == "list" {
+		memberType := shape.MemberRef.Shape.Type
+		return memberType == "structure" || memberType == "list" || memberType == "map"
+	}
+	return false
+}
+
+// collectSubShapes takes a shape and returns all the nested member and value
+// shapes that needs a helper function.
+func collectSubShapes(shape *awssdkmodel.Shape) []*awssdkmodel.Shape {
+	shapes := make([]*awssdkmodel.Shape, 0, 0)
+	if expectHelperFunction(shape) {
+		shapes = append(shapes, shape)
+	}
+
+	switch {
+	case shape.Type == "map":
+		valueShape := shape.ValueRef.Shape
+		switch valueShape.Type {
+		case "map":
+			valueShape := valueShape.ValueRef.Shape
+			shapes = append(shapes, collectSubShapes(valueShape)...)
+		case "list":
+			memberShape := valueShape.MemberRef.Shape
+			shapes = append(shapes, collectSubShapes(memberShape)...)
+		case "structure":
+			for _, memberName := range valueShape.MemberNames() {
+				memberShape := valueShape.MemberRefs[memberName].Shape
+				shapes = append(shapes, collectSubShapes(memberShape)...)
+			}
+		}
+	case shape.Type == "list":
+		memberShape := shape.MemberRef.Shape
+		switch memberShape.Type {
+		case "map":
+			valueShape := memberShape.ValueRef.Shape
+			shapes = append(shapes, collectSubShapes(valueShape)...)
+		case "list":
+			memberShape := memberShape.MemberRef.Shape
+			shapes = append(shapes, collectSubShapes(memberShape)...)
+		case "structure":
+			for _, memberName := range memberShape.MemberNames() {
+				memberShape := memberShape.MemberRefs[memberName].Shape
+				shapes = append(shapes, collectSubShapes(memberShape)...)
+			}
+		}
+	case shape.Type == "structure":
+		for _, memberName := range shape.MemberNames() {
+			memberShapeRef := shape.MemberRefs[memberName]
+			memberShape := memberShapeRef.Shape
+			shapes = append(shapes, collectSubShapes(memberShape)...)
+		}
+	}
+
+	return shapes
+}
+
+// equalMap outputs Go code that compares two Go maps of the the same value type.
+// at the first spotted difference the code return false, return true if the maps
+// key/values are equal.
+//
+// Output code will look something like this:
+//
+//  if len(a) != len(b) {
+//  	return false
+//  }
+//  for ka := range a {
+//  	_, ok := b[ka]
+//  	if !ok {
+//  		return false
+//  	}
+//  }
```

Comment:
nit: Remove references to iterators in code comments

Reply:
Yup this comment section is a little bit too large. Fixed